### ISSUE TITLE
Writing to a static is somehow allowed while initializing that same static

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust#142404]
I tried this code:

```rust
#![allow(dead_code)]

struct Foo {
    x: i32,
    y: (),
}

static S: Foo = Foo {
    x: 0,
    y: unsafe {
        (&raw const S.x).cast_mut().write(1);  // Writing to an immutable static that's still being initialized!
    },
};

fn main() {
    println!("{}", S.x);
}
```

I expected to see this happen: An error of some sort, or Miri reports UB.

Instead, this happened: The program compiles and outputs `0`. Miri does not detect UB.

Note that writing to a static while initializing a *different* static is not allowed. For example,

```rust
static mut S: i32 = 0;

static T: () = unsafe {
    (&raw mut S).write(1);
};
```

this produces the following error

```
error[E0080]: could not evaluate static initializer
    --> src/lib.rs:4:5
     |
4    |     (&raw mut S).write(1);
     |     ^^^^^^^^^^^^^^^^^^^^^ modifying a static's initial value from another static's initializer
     |
note: inside `std::ptr::mut_ptr::<impl *mut i32>::write`
    --> /playground/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mut_ptr.rs:1507:18
     |
1507 |         unsafe { write(self, val) }
     |                  ^^^^^^^^^^^^^^^^
note: inside `std::ptr::write::<i32>`
    --> /playground/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:1580:9
     |
1580 |         intrinsics::write_via_move(dst, src)
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the failure occurred here

For more information about this error, try `rustc --explain E0080`.
error: could not compile `playground` (lib) due to 1 previous error
```

### Meta

Reproducible on the playground with `1.89.0-nightly (2025-06-11 e703dff8fe220b78195c)`

@rustbot labels +A-const-eval

**Comment by theemathas** [comment]
Another variant:

```rust
#![allow(dead_code)]

struct Foo {
    x: i32,
    y: (),
}

static mut S: Foo = Foo {
    x: 0,
    y: unsafe {
        S.x = 1;
    },
};

fn main() {
    println!("{}", unsafe { S.x });
}
```

This code compiles and prints `0`.

**Comment by oli-obk** [comment]
Fun. we even have tests for this... but apparently something is not catching it

**Comment by oli-obk** [comment]
We thought https://github.com/rust-lang/rust/pull/116564 only made it possible to read from statics while initializing them (which just gives you uninit memory). But looking at the diff a test definitely changed to allow writing to statics, too

**Comment by theemathas** [comment]
From my testing, reading from a static while initializing it produces a "encountered static that tried to initialize itself with itself" error.

**Comment by oli-obk** [comment]
oh right, we had to prevent that for zsts as zsts perform no read, so we need to ensure they don't self-initialize

**Comment by oli-obk** [comment]
Or maybe we said this was fine since the value was always getting overwritten by the finial expression anyway? Because we had no way to allow the final write to the static while forbidding other writes. Slowly some memories of thinking about this before are coming up

**Comment by oli-obk** [comment]
Ok, yea, this was def intended (https://github.com/rust-lang/rust/pull/116564#issuecomment-1764302700), although arguably not discussed sufficiently. 

I had another look at the impl and it would need a bit of extra interpreter machine logic to prevent (or more specifically to special case `_0 = foo` MIR assignments to be ok, but all other uses of the static will get an error).

idk, considering 

> this was fine since the value was always getting overwritten by the finial expression anyway

There's nothing really important to change here. Considering

```rust
    let mut data = (0, 0);
    data = (1, {
        data.0 = 5;
        2
    });
    assert_eq!(data, (1, 2));
```

is fine, I don't see a good argument for rejecting the static equivalent

**Comment by Nemo157** [comment]
Might be a good candidate for a clippy lint in that case.

**Comment by oli-obk** [comment]
cc @rust-lang/wg-const-eval 

I think this is working as intended, but it is a bit surprising as it's rare that you can refer to the address of something before it is initialized

**Comment by RalfJung** [comment]
> (or more specifically to special case _0 = foo MIR assignments to be ok, but all other uses of the static will get an error)

IMO that is the main point: there's no difference between writing to `&raw mut S` and writing to `_0`, we just don't have syntax for the latter. `_0` always gets fully overwritten at the end, so these intermediate writes can't even have any effect. But semantically this must already work, so we may as well have it work in all cases.

But ultimately it is up to @rust-lang/lang whether they consider this problematic.

**Comment by lcnr** [comment]
I feel like I don't have a good tell on this behavior yet. It feels somewhat similar to the following, forbidden code:
```rust
#![allow(dead_code)]

struct Foo {
    x: i32,
    y: (),
}

fn main() {
    let s: Foo;
    s = Foo {
        x: 0,
        y: unsafe {
            (&raw const s.x).cast_mut().write(1); // Writing to an immutable static that's still being initialized!
        },
    };
    println!("{}", s.x);
}
```
If I understand correctly, we do not allow reading from statics in their initializer, i.e. the following errors with "could not evaluate static initializer" https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=47e4a7aebe19fdd8b8c7f4e45fa8fa81:
```rust
static S: Foo = Foo {
    x: 0,
    y: unsafe {
        let _ = (&raw const S.x).cast::<()>().read(); 
    },
};
```
Is the following summary correct:
- reading from statics in statics initializers is required to be non-cyclic. you can't read from the static you're currently initializing
- writing to the (regardless of mutability) static we're currently initializing is allowed and will always get overwritten by the final write to the return place of the initializer

This means while the writes to the static are weird, they do not cause any opsem or soundness issues.

I personally feel like we should forbid these useless writes. Afaict they are always irrelevant and I can't think of a case where it would be desirable. If we only allow initializing the static with the final value of its initializer were trivial to do, do you think we'd still want to support these irrelevant earlier writes?

**Comment by oli-obk** [comment]
> I personally feel like we should forbid these useless writes. Afaict they are always irrelevant and I can't think of a case where it would be desirable. If we only allow initializing the static with the final value of its initializer were trivial to do, do you think we'd still want to support these irrelevant earlier writes?

I don't want to support these irrelevant earlier writes. It's just a bit of extra complexity to prevent them, so we actually need to want to prevent them explicitly and thus take the complexity hit.

**Comment by RalfJung** [comment]
I wouldn't mind forbidding them if there was a nice way to do it. But syntactically special-casing the `_0 = expr` write is a bit too dirt of a hack IMO.

**Comment by lcnr** [comment]
It feels somewhat surprising to me that we even treat the static and the return place as the same thing instead of evaluating the static initializer just like another const body and then manually using its result as the value of the static 🤔 why is this the case?

**Comment by RalfJung** [comment]
We do treat it like any other const body. Const bodies evaluate into a return destination (i.e., they return a place, not a value). For statics, we make that return destination be the static itself.

For some time we had a "temporary" return destination, but that lead to the uncomfortable situation where the AllocId representing the static was different when evaluating the static vs later.

**Comment by lcnr** [comment]
> but that lead to the uncomfortable situation where the AllocId representing the static was different when evaluating the static vs later.

why this? and how do we make sure that the `AllocId` used for other statics doesn't differ between statics we evaluate?

**Comment by lcnr** [comment]
I guess we also have `AllocId`s for locals?

**Comment by RalfJung** [comment]
> why this? and how do we make sure that the AllocId used for other statics doesn't differ between statics we evaluate?

We don't have two AllocIds for a static any more so there's no problem any more. :)

And yes, every allocation has an AllocId -- locals, globals, heap allocations, vtables, functions (as the target of function pointers), everything.

**Comment by bjorn3** [comment]
> IMO that is the main point: there's no difference between writing to &raw mut S and writing to _0, we just don't have syntax for the latter. _0 always gets fully overwritten at the end, so these intermediate writes can't even have any effect. But semantically this must already work, so we may as well have it work in all cases.

If we allow this to work, then we won't be able to do RVO anymore, nor would language builtin pin-init be possible. As those would avoid _0 being fully overwritten after `(&raw const S.x).cast_mut().write(1)` runs.

**Comment by RalfJung** [comment]
We don't optimize consts so RVO isn't much of a concern. Not sure how pin-init would work in consts?

**Comment by traviscross** [comment]
> But ultimately it is up to [@rust-lang/lang](https://github.com/orgs/rust-lang/teams/lang) whether they consider this problematic.

Thanks for the report and analysis.  We'll discuss.

For my part, as @lcnr pointed out, it does seem a bit odd that these are inconsistent:

```rust
// With local.
let x: (u8, ());
x = (0, unsafe {
    (&raw const x.0).cast_mut().write(1);
    //~^ ERROR used binding `x` isn't initialized
});
// With static.
static X: (u8, ()) = (0, unsafe {
    (&raw const X.0).cast_mut().write(1);
    //~^ OK?
});
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&code=%2F%2F+This+demonstrates+the+issue+raised+in+142404+about+being+able+to%0A%2F%2F+write+to+a+static+in+its+initializer+without+the+language+or+Miri%0A%2F%2F+complaining.%0A%2F%2F%0A%2F%2F+See%3A%0A%2F%2F%0A%2F%2F+-+https%3A%2F%2Fgithub.com%2Frust-lang%2Frust%2Fissues%2F142404%0A%2F%2F%0A%2F%2F+Author%3A+TC%0A%2F%2F+Date%3A+2025-06-13%0A%0Afn+main%28%29+%7B%0A++++%2F%2F+With+local.%0A++++let+x%3A+%28u8%2C+%28%29%29%3B%0A++++x+%3D+%280%2C+unsafe+%7B%0A++++++++%28%26raw+const+x.0%29.cast_mut%28%29.write%281%29%3B%0A++++++++%2F%2F~%5E+ERROR+used+binding+%60x%60+isn%27t+initialized%0A++++%7D%29%3B%0A++++assert_eq%21%28x.0%2C+0%29%3B%0A++++%2F%2F+With+static.%0A++++static+X%3A+%28u8%2C+%28%29%29+%3D+%280%2C+unsafe+%7B%0A++++++++%28%26raw+const+X.0%29.cast_mut%28%29.write%281%29%3B%0A++++++++%2F%2F~%5E+OK%3F%0A++++%7D%29%3B%0A++++assert_eq%21%28X.0%2C+0%29%3B%0A%7D%0A)

Regarding the complexity hit here, sometimes, in the past, it's been easier to add a lint for something than to make it a hard error (since it was OK if the lint was only mostly correct).  Any chance that this is one of those cases?

**Comment by RalfJung** [comment]
> it does seem a bit odd that these are inconsistent:

This is a direct consequence of the following, fully intentional "inconsistency":
```rust
struct S(&'static S);

static MY_S: S = S(&MY_S); // works

let my_s = S(&my_s); // does not work
```

**Comment by traviscross** [comment]
> This is a direct consequence of the following, fully intentional "inconsistency":
>
> ```rust
> struct S(&'static S);
> static MY_S: &'static S = S(&MY_S); // works
> let my_s = S(&my_s); // does not work
> ```

I get:

```rust
struct S(&'static S);
static MY_S: &'static S = &S(&MY_S);
//~^ ERROR could not evaluate static initializer
//~| ERROR encountered static that tried to initialize itself with itself
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&code=struct+S%28%26%27static+S%29%3B%0Astatic+MY_S%3A+%26%27static+S+%3D+%26S%28%26MY_S%29%3B%0A%2F%2F~%5E+ERROR+could+not+evaluate+static+initializer%0A%2F%2F~%7C+ERROR+encountered+static+that+tried+to+initialize+itself+with+itself%0A%0Afn+main%28%29+%7B%7D%0A)

**Comment by RalfJung** [comment]
Sorry, I had a little typo:
```rust
struct S(&'static S);
static MY_S: S = S(&MY_S); // works
```

**Comment by traviscross** [comment]
Thanks.  Is the following also equivalent to your point (i.e. without that `'static` lifetime)?:

```rust
struct S(*const S);
unsafe impl Sync for S {}

fn main() {
    static MY_S: S = S(&raw const MY_S);
    let my_s;
    my_s = S(&raw const my_s);
    //~^ ERROR used binding `my_s` isn't initialized
}
```

**Comment by RalfJung** [comment]
> Thanks. Is the following also equivalent to your point 

Yes. More broadly, we allow statics to recursively point to themselves. This goes waaaay back. That makes statics consistent with other items like functions and types that are also allowed to recursively refer to themselves.

We don't allow the same for `let` bindings. I don't know why this choice was made, but it is the common choice among languages I know of. It is also crucial to allow shadowing patterns like `let x = convert(x);`.

**Comment by traviscross** [comment]
I'd expect `let x; x = ..` and `let x = ..` to not have the same behavior with respect to shadowing.

**Comment by RalfJung** [comment]
Yeah, if we allowed taking a raw pointer to an uninitialized local we would likely accept that code. We'd have to ensure the local is considered "live" by MIR building the moment a raw borrow occurs.

**Comment by theemathas** [comment]
Interestingly, this outputs `1`.

```rust
#![feature(const_destruct, const_trait_impl)]

struct Weird;
impl const Drop for Weird {
    fn drop(&mut self) {
        unsafe { X = 2 };
    }
}

static mut X: i32 = (1, Weird).0;

fn main() {
    println!("{}", unsafe { X });
}
```

The MIR seems to suggest that the `X = 2` write occurs after writing `1` to `_0` though, which confuses me.

<details>
<summary>MIR</summary>

```
// MIR FOR CTFE
fn <impl at [src/main.rs:4:1: 4:26](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024#)>::drop(_1: &mut Weird) -> () {
    debug self => _1;
    let mut _0: ();
    let _2: ();
    let mut _3: *mut i32;

    bb0: {
        StorageLive(_2);
        StorageLive(_3);
        _3 = const {alloc1: *mut i32};
        (*_3) = const 2_i32;
        _2 = const ();
        StorageDead(_3);
        StorageDead(_2);
        _0 = const ();
        return;
    }
}

alloc1 (static: X)

static mut X: i32 = {
    let mut _0: i32;
    let mut _1: (i32, Weird);
    let mut _2: Weird;

    bb0: {
        StorageLive(_1);
        StorageLive(_2);
        _2 = Weird;
        _1 = (const 1_i32, move _2);
        StorageDead(_2);
        _0 = copy (_1.0: i32);
        drop(_1) -> [return: bb1, unwind: bb2];
    }

    bb1: {
        StorageDead(_1);
        return;
    }

    bb2 (cleanup): {
        resume;
    }
}
```

</details>

**Comment by RalfJung** [comment]
Nice example! I suspect `_0` in the MIR is not the same as `X`. After the body is done evaluating, we essentially copy the contents of `_0` over to `X`.

Hm, maybe that means we *can* prevent these writes? We just have to "turn off" that guard before we do the final copy.

**Comment by RalfJung** [comment]
Yeah we create a separate return place here

https://github.com/rust-lang/rust/blob/17946c22b1d7abd2dd990bf6998c8491b534fe62/compiler/rustc_const_eval/src/const_eval/eval_queries.rs#L61-L65

So, we should actually be able to block all writes to that place until just before we execute the `Return` terminator of the initializer expression. We can use the `before_stack_pop` to detect that this is about to happen.

**Comment by oli-obk** [comment]
well, yes we copy the _0 local var into the return place when returning from a function,

https://github.com/rust-lang/rust/blob/8bce2255e826a11d1aa345d1786f85d22c5f921a/compiler/rustc_const_eval/src/interpret/call.rs#L903

but how much is that actually part of opsem? Wouldn't overwriting the return place with uninit before calling the function also be a legal implementation with effectively the same semantics except that you can possibly observe partial writes that happened before the function unwinds?

**Comment by RalfJung** [comment]
The return place is already uninit before calling the function, and we anyway have no syntax to access the return place, so I don't understand the question.

**Comment by oli-obk** [comment]
Basically I'm asking why we even store things in `_0` and make it not just refer again to the return place ("again" because I'm fairly certain we used to have that and I thought we only switched to the copying model because it was somehow observable)

**Comment by RalfJung** [comment]
Well, the copying model might make it possible to reject code like the initial example here, which would be a good reason to keep it IMO.

Also conceptually, `_0` is a local variable in the callee and the caller separately gives its own return place -- I don't think we should mix those up.
